### PR TITLE
Restructure and improve docs in apps.train module

### PR DESCRIPTION
**Pull request by trbromley**
This PR does the following:

- Changes the import style so that key classes are available directly within `apps.train`
- Collates the `embed`, `param`, and `cost` module docstrings into the `apps.train` docstring

### CODE REVIEW

File: strawberryfields/apps/train/__init__.py [hunk 1]
```diff
@@ -3,19 +3,87 @@
...
+Training algorithms for GBS distributions rely on the :math:`WAW` parametrization, where :math:`W`
+is a diagonal matrix of weights and :math:`A` is a symmetric matrix. Trainable parameters are
+embedded into the GBS distribution by expressing the weights as functions of the parameters.
+
+This module contains methods to implement such embeddings. It also provides derivatives
+of the weights with respect to the trainable parameters. There are two main classes, each
+corresponding to a different embedding. The :class:`Exp` class is a simple embedding where the
+weights are exponentials of the trainable parameters. The :class:`ExpFeatures` class is a more
+general embedding that makes use of user-defined feature vectors, which potentially provide more
+flexibility in training strategies.
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:23]: This is taken from the `embed` module.

File: strawberryfields/apps/train/__init__.py [hunk 2]
```diff
@@ -3,19 +3,87 @@
 
 .. currentmodule:: strawberryfields.apps.train
 
+A GBS device can be treated as a variational quantum device with trainable squeezing,
+beamsplitter and rotation gates. This module provides the functionality for training a GBS device
+using the approach outlined in `this <https://arxiv.org/abs/2004.04770>`__ paper
+:cite:`banchi2020training`.
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:9]: This is a new introduction.

File: strawberryfields/apps/train/__init__.py [hunk 3]
```diff
@@ -3,19 +3,87 @@
...
+
+Additional functionality related to the variational GBS model is available.
+
+.. autosummary::
+    :toctree: api
+
+    A_to_cov
+    prob_click
+    prob_photon_sample
+    rescale_adjacency
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:63]: This is new.

File: strawberryfields/apps/train/__init__.py [hunk 4]
```diff
@@ -3,19 +3,87 @@
...
+
+Choosing a cost function
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+In the context of stochastic optimization, cost functions are expressed as expectation values
+over the GBS distribution. Within the WAW parametrization, gradients of cost functions can also be
+expressed as expectation values over the GBS distribution. This module contains methods for
+calculating these gradients and for using gradient-based methods to optimize GBS circuits. In the
+case of optimization with respect to a Kullback-Leibler divergence or log-likelihood cost
+function, gradients can be computed efficiently, leading to fast training.
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:73]: This is taken from the `cost` module.

File: strawberryfields/apps/train/__init__.py (hunk 1 above)
Comment (trbromley) [reply strawberryfields/apps/train/__init__.py:23]: We may come back to change wording here if needed, but to keep this PR light I have just copy/paste (+ a minor change due to the move).

File: strawberryfields/apps/train/__init__.py (hunk 4 above)
Comment (trbromley) [reply strawberryfields/apps/train/__init__.py:73]: We may come back to change wording here if needed, but to keep this PR light I have just copy/paste (+ a minor change due to the move).

File: strawberryfields/apps/train/__init__.py [hunk 5]
```diff
@@ -3,19 +3,87 @@
...
-from strawberryfields.apps.train.cost import Stochastic
+from strawberryfields.apps.train.cost import KL, Stochastic
+from strawberryfields.apps.train.embed import Exp, ExpFeatures
+from strawberryfields.apps.train.param import (
+    VGBS,
+    A_to_cov,
+    prob_click,
+    prob_photon_sample,
+    rescale_adjacency,
+)
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:89]: Note this part - we no longer make visible the submodules to users. Instead, all relevant parts are imported here and the use of submodules is just for book keeping.

File: strawberryfields/apps/train/__init__.py [hunk 6]
```diff
@@ -3,19 +3,87 @@
 
 .. currentmodule:: strawberryfields.apps.train
 
+A GBS device can be treated as a variational quantum device with trainable squeezing,
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:6]: The easiest way to visualise the changes is by going here:
https://strawberryfields--414.org.readthedocs.build/en/414/code/api/strawberryfields.apps.train.html#module-strawberryfields.apps.train

File: doc/references.bib [hunk 7]
```diff
@@ -1025,3 +1025,10 @@ @article{jacob2009localizing
 pages = {084106},
 year = {2009}
 }
+
+@article{banchi2020training,
+  title={Training Gaussian Boson Sampling Distributions},
+  author={Banchi, Leonardo and Quesada, Nicol{\'a}s and Arrazola, Juan Miguel},
+  journal={arXiv preprint arXiv:2004.04770},
```
Comment (ixfoduap) [inline doc/references.bib:1032]:
```suggestion
  journal={arXiv:2004.04770},
```

File: strawberryfields/apps/train/__init__.py (hunk 6 above)
Comment (ixfoduap) [reply strawberryfields/apps/train/__init__.py:6]: This is so helpful Tom! :mage:

File: strawberryfields/apps/train/__init__.py [hunk 8]
```diff
@@ -3,19 +3,87 @@
...
+beamsplitter and rotation gates. This module provides the functionality for training a GBS device
+using the approach outlined in `this <https://arxiv.org/abs/2004.04770>`__ paper
```
Comment (ixfoduap) [inline strawberryfields/apps/train/__init__.py:8]:
```suggestion
using the approach outlined in `this paper <https://arxiv.org/abs/2004.04770>`__ 
```

File: strawberryfields/apps/train/__init__.py (hunk 1 above)
Comment (ixfoduap) [reply strawberryfields/apps/train/__init__.py:23]: I think it read well and could be left as is

File: strawberryfields/apps/train/__init__.py (hunk 4 above)
Comment (ixfoduap) [reply strawberryfields/apps/train/__init__.py:73]: Again, it reads well and I think we can keep as is

File: strawberryfields/apps/train/__init__.py (hunk 5 above)
Comment (ixfoduap) [reply strawberryfields/apps/train/__init__.py:89]: This is amazing!

File: strawberryfields/apps/train/__init__.py [hunk 9]
```diff
@@ -3,19 +3,87 @@
 
 .. currentmodule:: strawberryfields.apps.train
 
+A GBS device can be treated as a variational quantum device with trainable squeezing,
+beamsplitter and rotation gates. This module provides the functionality for training a GBS device
+using the approach outlined in `this paper <https://arxiv.org/abs/2004.04770>`__
```
Comment (trbromley) [inline strawberryfields/apps/train/__init__.py:8]:
```suggestion
using the approach outlined in `this paper <https://arxiv.org/abs/2004.04770>`__
```

**Pull request merged**